File: control-contest-02-prepare/B/python/B.py
import sys

input = sys.stdin.readline


class DSU:
    def __init__(self, n):
        self.parent = list(range(n))
        self.size = [1] * n

    def get(self, v):
        while self.parent[v] != v:
            self.parent[v] = self.parent[self.parent[v]]
            v = self.parent[v]
        return v

    def union(self, a, b):
        a = self.get(a)
        b = self.get(b)

        if a == b:
            return False

        if self.size[a] < self.size[b]:
            a, b = b, a

        self.parent[b] = a
        self.size[a] += self.size[b]
        return True


n = int(input())

weights = [list(map(int, input().split())) for _ in range(n)]

dsu = DSU(n)
edges = []

for i in range(n):
    row = list(map(int, input().split()))

    for j in range(i + 1, n):
        if row[j] == 1:
            dsu.union(i, j)
        else:
            edges.append((weights[i][j], i, j))

edges.sort()

answer = 0

for w, u, v in edges:
    if dsu.union(u, v):
        answer += w

print(answer)

File: control-contest-02-prepare/B/java/B.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class B {
    static class DSU {
        int[] parent;
        int[] size;

        DSU(int n) {
            parent = new int[n];
            size = new int[n];

            for (int i = 0; i < n; i++) {
                parent[i] = i;
                size[i] = 1;
            }
        }

        int get(int v) {
            while (parent[v] != v) {
                parent[v] = parent[parent[v]];
                v = parent[v];
            }
            return v;
        }

        boolean union(int a, int b) {
            a = get(a);
            b = get(b);

            if (a == b) {
                return false;
            }

            if (size[a] < size[b]) {
                int temp = a;
                a = b;
                b = temp;
            }

            parent[b] = a;
            size[a] += size[b];
            return true;
        }
    }

    static class Edge implements Comparable<Edge> {
        int weight;
        int from;
        int to;

        Edge(int weight, int from, int to) {
            this.weight = weight;
            this.from = from;
            this.to = to;
        }

        @Override
        public int compareTo(Edge other) {
            return Integer.compare(this.weight, other.weight);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(in.readLine());

        int[][] weights = new int[n][n];

        for (int i = 0; i < n; i++) {
            String[] parts = in.readLine().split(" ");
            for (int j = 0; j < n; j++) {
                weights[i][j] = Integer.parseInt(parts[j]);
            }
        }

        DSU dsu = new DSU(n);
        List<Edge> edges = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            String[] parts = in.readLine().split(" ");

            for (int j = i + 1; j < n; j++) {
                int hasBridge = Integer.parseInt(parts[j]);

                if (hasBridge == 1) {
                    dsu.union(i, j);
                } else {
                    edges.add(new Edge(weights[i][j], i, j));
                }
            }
        }

        Collections.sort(edges);

        int answer = 0;

        for (Edge edge : edges) {
            if (dsu.union(edge.from, edge.to)) {
                answer += edge.weight;
            }
        }

        System.out.println(answer);
    }
}